components/camera/replaypcd/replaypcd_test.go

```diff
@@ -2,27 +2,404 @@ package replaypcd
+				},
+			},
+			startFileNum: 0,
+			endFileNum:   10,
+		},
+		{
+			description: "Calling NextPointCloud with start filter",
+			cfg: &Config{
+				Source: "source",
+				Interval: TimeInterval{
+					Start: "2000-01-01T12:00:05Z",
+				},
+			},
+			startFileNum: 5,
+			endFileNum:   numPCDFiles,
+		},
+		{
+			description: "Calling NextPointCloud with start and end filter",
+			cfg: &Config{
+				Source: "source",
+				Interval: TimeInterval{
+					Start: "2000-01-01T12:00:05Z",
+					End:   "2000-01-01T12:00:10Z",
+				},
+			},
+			startFileNum: 5,
+			endFileNum:   10,
+		},
+		{
+			description: "Calling NextPointCloud with bad source",
+			cfg: &Config{
+				Source: "bad_source",
+			},
+			startFileNum: -1,
+			endFileNum:   -1,
+		},
+		{
+			description: "Calling NextPointCloud with robot_id",
+			cfg: &Config{
+				Source:  "source",
+				RobotID: "robot_id",
+			},
+			startFileNum: 0,
+			endFileNum:   numPCDFiles,
+		},
+		{
+			description: "Calling NextPointCloud with bad robot_id",
+			cfg: &Config{
+				Source:  "source",
+				RobotID: "bad_robot_id",
+			},
+			startFileNum: -1,
+			endFileNum:   -1,
+		},
+	}
+
+	for _, tt := range cases {
+		t.Run(tt.description, func(t *testing.T) {
+			replayCamera, serverClose, err := createNewReplayPCDCamera(ctx, t, tt.cfg, true)
+			test.That(t, err, test.ShouldBeNil)
+			test.That(t, replayCamera, test.ShouldNotBeNil)
+
+			// Iterate through all files that meet the provided filter
+			if tt.startFileNum != -1 {
+				for i := tt.startFileNum; i < tt.endFileNum; i++ {
+					pc, err := replayCamera.NextPointCloud(ctx)
+					test.That(t, err, test.ShouldBeNil)
+					test.That(t, pc, test.ShouldResemble, getPointCloudFromArtifact(t, i))
+				}
+			}
+
+			// Confirm the end of the dataset was reached when expected
+			pc, err := replayCamera.NextPointCloud(ctx)
+			test.That(t, err, test.ShouldNotBeNil)
+			test.That(t, err.Error(), test.ShouldContainSubstring, errEndOfDataset.Error())
+			test.That(t, pc, test.ShouldBeNil)
+
+			err = replayCamera.Close(ctx)
+			test.That(t, err, test.ShouldBeNil)
+
+			test.That(t, serverClose(), test.ShouldBeNil)
+		})
+	}
+}
+
+func TestLiveNextPointCloud(t *testing.T) {
 	ctx := context.Background()
 
-	cfg := resource.Config{}
+	numPCDFilesOriginal := numPCDFiles
+	numPCDFiles = 10
+	defer func() { numPCDFiles = numPCDFilesOriginal }()
 
-	replayCamera, err := newPCDCamera(ctx, nil, cfg, logger)
+	cfg := &Config{
+		Source: "source",
+	}
+
+	replayCamera, serverClose, err := createNewReplayPCDCamera(ctx, t, cfg, true)
 	test.That(t, err, test.ShouldBeNil)
+	test.That(t, replayCamera, test.ShouldNotBeNil)
 
-	t.Run("NextPointCloud", func(t *testing.T) {
-		_, err := replayCamera.NextPointCloud(ctx)
-		test.That(t, err.Error(), test.ShouldEqual, "NextPointCloud is unimplemented")
-	})
+	// Iterate through all files that meet the provided filter
+	i := 0
+	for {
+		pc, err := replayCamera.NextPointCloud(ctx)
+		if i >= numPCDFiles {
```

Comment:
Would `i == numPCDFiles` be sufficient? The >= condition threw me off.

Reply:
I think its best practice to make checks like these as robust as possible. Same reason in a for loop we do:

for(i = 0; i < size; i ++) {}

If anything else in the loop were to be added that might change the value of I we want to ensure failure as accessing an element (or in this case an artifact) above the defined size should/would cause an error

Reply:
Spoke with Jeremy offline, I agree that safety is important for production code paths but since this is a test I'd prefer that we do the minimum required check here. It makes the test easier to grok and if someone modifies this code such that we can get `i > numPCDFiles`, I'd rather we hit an error getting the file so the developer is aware that they messed something up.